### PR TITLE
GDScript: Optimize non-constant `for`-`range`

### LINKED ISSUE #83293: for loop with large int values returns negative numbers
### Godot version

Godot v4.1.1.stable

### System information

Godot v4.1.1.stable - Windows 10.0.19045 - Vulkan (Forward+) - dedicated GeForce GT 640 () - AMD FX(tm)-8320 Eight-Core Processor (8 Threads)

### Issue description

`for` loops with large `int` values return negative values instead of the correct values.

### Steps to reproduce

Example code:

```
const BLICKS:int = 705600000
var start_loop:int = BLICKS * 40
var end_loop:int = BLICKS * 42
var step_loop: int = BLICKS / 32
for i in range( start_loop, end_loop, step_loop ):
		print(i)
```

Example output:
```

-1840771072
-1818721072
-1796671072
-1774621072
-1752571072
-1730521072
-1708471072
...
```

In contrast, a `while` loop returns the correct values:
```
const BLICKS:int = 705600000
var start_loop:int = BLICKS * 40
var end_loop:int = BLICKS * 42
var step_loop: int = BLICKS / 32
while start_loop < end_loop:
	print(start_loop)
	start_loop += step_loop
```
Correct output:
```
28224000000
28246050000
28268100000
28290150000
28312200000
28334250000
28356300000
28378350000
...
```

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by dalexeev**
* Fixes #83293.

### CODE REVIEW

**Comment by dalexeev** [comment]
Is there a way to improve error messages?

```gdscript
var a: Variant = 1
var b: Variant = 8
var c: Variant = 2
var d: Variant = []

for i in range(a, b, c): # 1, 3, 5, 7
    print(var_to_str(i))

for i in range(a, b, d): # Runtime error: Invalid call. Nonexistent 'Vector3i' constructor.
    print(var_to_str(i))
```

File: modules/gdscript/gdscript_analyzer.cpp [hunk 1]
```diff
@@ -1700,19 +1700,33 @@ void GDScriptAnalyzer::resolve_for(GDScriptParser::ForNode *p_for) {
...
-				if (p_for->list->is_constant) {
-					p_for->list->set_datatype(type_from_variant(p_for->list->reduced_value, p_for->list));
-				} else {
-					GDScriptParser::DataType list_type;
-					list_type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
-					list_type.kind = GDScriptParser::DataType::BUILTIN;
-					list_type.builtin_type = Variant::ARRAY;
-					p_for->list->set_datatype(list_type);
+						call->function_name = Variant::get_type_name(range_replacement);
+						callee->name = Variant::get_type_name(range_replacement);
```
Comment (vnen) [inline modules/gdscript/gdscript_analyzer.cpp:1722]: I'm not really fond of changing what's in the parser tree like this. Mostly because if you try to print the tree afterwards (which I do for debugging) you'll get a call to `Vector2i` instead of `range` which might get confusing, even if you know about this.

Comment (vnen) [reply modules/gdscript/gdscript_analyzer.cpp:1722]: The proper solution would be to flag this call so the compiler can emit the instruction for constructor instead of a call to `range`.

**Comment by Calinou** [comment]
Can you provide before/after benchmarks with a release export template?

File: modules/gdscript/gdscript_analyzer.cpp [hunk 2]
```diff
@@ -1700,19 +1700,33 @@ void GDScriptAnalyzer::resolve_for(GDScriptParser::ForNode *p_for) {
...
+							case 1:
+								range_replacement = Variant::INT;
+								break;
+							case 2:
+								range_replacement = Variant::VECTOR2I;
+								break;
+							case 3:
+								range_replacement = Variant::VECTOR3I;
+								break;
+						}
```
Comment (vnen) [inline modules/gdscript/gdscript_analyzer.cpp:1719]: One issue here is that technically floats are allowed, but this is forcing it to be int in all cases. This is tricky because there's no way to know until runtime. I wonder if this needs a special instruction in the VM so it can properly check the types (also would allow to have a more specific error message, as you asked).

**Comment by dalexeev** [comment]
I think we can defer this to 4.1 as it's not a blocker or a bug. And if someone has a problem allocating a large array, they can explicitly use `int`/`Vector2i`/`Vector3i` or replace `for` with `while` as a workaround.

File: modules/gdscript/gdscript_analyzer.cpp (hunk 2 above)
Comment (dalexeev) [reply modules/gdscript/gdscript_analyzer.cpp:1719]: I think we need `rangef` or `range_float` since `for` already supports `float`, `Vector2`, `Vector3`.

**Comment by dalexeev** [comment]
### Benchmarks

**Test script**

```gdscript
extends Node

func _ready() -> void:
    await get_tree().create_timer(3.0).timeout
    print("Test 1. Simple loop.")
    await get_tree().create_timer(3.0).timeout
    var a := 50_000_000
    var start_time := Time.get_ticks_msec()
    for i in range(0, a):
        var x := i
    var time := Time.get_ticks_msec() - start_time
    print("%d msec" % time)

    await get_tree().create_timer(3.0).timeout
    print("Test 2. Nested loop.")
    await get_tree().create_timer(3.0).timeout
    a = 10_000
    start_time = Time.get_ticks_msec()
    for i in range(0, a):
        for j in range(i, a):
            var x := j
    time = Time.get_ticks_msec() - start_time
    print("%d msec" % time)
```

**Before (`dev_build=true tests=true`)**

```
Test 1. Simple loop. # ~1 GiB RAM.
7753 msec
Test 2. Nested loop.
8010 msec
```

**Before (`target=template_release`)**

```
Test 1. Simple loop. # ~1 GiB RAM.
1568 msec
Test 2. Nested loop.
1404 msec
```


**After (`dev_build=true tests=true`)**

```
Test 1. Simple loop. # ~0 GiB RAM.
2934 msec
Test 2. Nested loop.
2941 msec
```

**After (`target=template_release`)**

```
Test 1. Simple loop. # ~0 GiB RAM.
426 msec
Test 2. Nested loop.
433 msec
```

File: modules/gdscript/gdscript_vm.cpp [hunk 3]
```diff
@@ -1591,6 +1592,59 @@ Variant GDScriptFunction::call(GDScriptInstance *p_instance, const Variant **p_a
...
+					case 1:
+						list_type = Variant::INT;
+						break;
+					case 2:
+						list_type = Variant::VECTOR2I;
+						break;
+					case 3:
+						list_type = Variant::VECTOR3I;
+						break;
+				}
```
Comment (dalexeev) [inline modules/gdscript/gdscript_vm.cpp:1614]: Should we make 3 separate opcodes?

**Comment by dalexeev** [comment]
### Updated benchmarks

<details>
<summary>Test script</summary>

```gdscript
extends Node

var _mem_before: float
var _mem_after: float

func simple_alt_to() -> void:
    var a := 50_000_000
    _mem_before = Performance.get_monitor(Performance.MEMORY_STATIC)
    for i in int(a):
        if i == 0:
            _mem_after = Performance.get_monitor(Performance.MEMORY_STATIC)
        var _x := i

func simple_alt_from_to() -> void:
    var a := 50_000_000
    _mem_before = Performance.get_monitor(Performance.MEMORY_STATIC)
    for i in Vector2i(0, a):
        if i == 0:
            _mem_after = Performance.get_monitor(Performance.MEMORY_STATIC)
        var _x := i

func simple_alt_from_to_step() -> void:
    var a := 50_000_000
    _mem_before = Performance.get_monitor(Performance.MEMORY_STATIC)
    for i in Vector3i(0, a, 1):
        if i == 0:
            _mem_after = Performance.get_monitor(Performance.MEMORY_STATIC)
        var _x := i

func simple_range_to() -> void:
    var a := 50_000_000
    _mem_before = Performance.get_monitor(Performance.MEMORY_STATIC)
    for i in range(a):
        if i == 0:
            _mem_after = Performance.get_monitor(Performance.MEMORY_STATIC)
        var _x := i

func simple_range_from_to() -> void:
    var a := 50_000_000
    _mem_before = Performance.get_monitor(Performance.MEMORY_STATIC)
    for i in range(0, a):
        if i == 0:
            _mem_after = Performance.get_monitor(Performance.MEMORY_STATIC)
        var _x := i

func simple_range_from_to_step() -> void:
    var a := 50_000_000
    _mem_before = Performance.get_monitor(Performance.MEMORY_STATIC)
    for i in range(0, a, 1):
        if i == 0:
            _mem_after = Performance.get_monitor(Performance.MEMORY_STATIC)
        var _x := i

func nested_range_to() -> void:
    var a := 50_000
    var b := 1000
    _mem_before = Performance.get_monitor(Performance.MEMORY_STATIC)
    for i in range(a):
        for j in range(b):
            if i == 0 and j == 0:
                _mem_after = Performance.get_monitor(Performance.MEMORY_STATIC)
            var _x := j

func nested_range_from_to() -> void:
    var a := 50_000
    var b := 1000
    _mem_before = Performance.get_monitor(Performance.MEMORY_STATIC)
    for i in range(0, a):
        for j in range(0, b):
            if i == 0 and j == 0:
                _mem_after = Performance.get_monitor(Performance.MEMORY_STATIC)
            var _x := j

func nested_range_from_to_step() -> void:
    var a := 50_000
    var b := 1000
    _mem_before = Performance.get_monitor(Performance.MEMORY_STATIC)
    for i in range(0, a, 1):
        for j in range(0, b, 1):
            if i == 0 and j == 0:
                _mem_after = Performance.get_monitor(Performance.MEMORY_STATIC)
            var _x := j

func benchmark(test_name: String, test_func: Callable) -> void:
    var start_time := Time.get_ticks_msec()
    test_func.call()
    var mem := _mem_after - _mem_before
    var time := Time.get_ticks_msec() - start_time
    print("%s - %s - %d msec" % [test_name, String.humanize_size(int(mem)), time])

func _ready() -> void:
    print("Please wait...")
    await get_tree().create_timer(3.0).timeout
    benchmark("Simple alt to", simple_alt_to)
    benchmark("Simple alt from-to", simple_alt_from_to)
    benchmark("Simple alt from-to-step", simple_alt_from_to_step)
    benchmark("Simple range to", simple_range_to)
    benchmark("Simple range from-to", simple_range_from_to)
    benchmark("Simple range from-to-step", simple_range_from_to_step)
    benchmark("Nested range to", nested_range_to)
    benchmark("Nested range from-to", nested_range_from_to)
    benchmark("Nested range from-to-step", nested_range_from_to_step)
    print("Done!")
```

</details>

(Monitor `Performance.MEMORY_STATIC` is not available in release builds.)

```
Before (dev_build=true tests=true)
==================================
Simple alt to - 0 B - 7306 msec
Simple alt from-to - 0 B - 7278 msec
Simple alt from-to-step - 0 B - 7243 msec
Simple range to - 2.00 GiB - 12619 msec
Simple range from-to - 2.00 GiB - 12396 msec
Simple range from-to-step - 2.00 GiB - 12544 msec
Nested range to - 2.03 MiB - 13774 msec
Nested range from-to - 2.03 MiB - 13782 msec
Nested range from-to-step - 2.03 MiB - 14062 msec

Before (target=template_release)
================================
Simple alt to - 843 msec
Simple alt from-to - 826 msec
Simple alt from-to-step - 835 msec
Simple range to - 2207 msec
Simple range from-to - 2206 msec
Simple range from-to-step - 2245 msec
Nested range to - 2329 msec
Nested range from-to - 2252 msec
Nested range from-to-step - 2251 msec

After (dev_build=true tests=true)
==================================
Simple alt to - 0 B - 7267 msec
Simple alt from-to - 0 B - 6984 msec
Simple alt from-to-step - 0 B - 7150 msec
Simple range to - 0 B - 7488 msec
Simple range from-to - 0 B - 7527 msec
Simple range from-to-step - 0 B - 7465 msec
Nested range to - 0 B - 9333 msec
Nested range from-to - 0 B - 9305 msec
Nested range from-to-step - 0 B - 9329 msec

After (target=template_release)
================================
Simple alt to - 872 msec
Simple alt from-to - 825 msec
Simple alt from-to-step - 828 msec
Simple range to - 860 msec
Simple range from-to - 912 msec
Simple range from-to-step - 907 msec
Nested range to - 1274 msec
Nested range from-to - 1276 msec
Nested range from-to-step - 1344 msec
```

**Review by vnen — APPROVED** [review]

**Comment by dalexeev** [comment]
Rebased just in case.

**Comment by Repiteo** [comment]
Thanks!

**Comment by Macksaur** [comment]
Heya, this PR broke some of the nullable static analysis I was doing as it's deleted some constant reductions. I was relying on the following being set as part of reduction:
```cpp
p_for->list->is_constant = true;
p_for->list->reduced_value = reduced;
```
I was using it to identify some edge cases when variables aren't nullable in certain deducible constant contexts:
```gdscript
func check(n:int) -> void:
	print(n)

func test():
	var foo:int?
	for i in range(0, 1):
		foo = 1
	check(foo) # foo is always non-null
```
`>> ERROR at line 8: Invalid argument for "check()" function: argument 1 must not be null, but "int?" is nullable.`

Are there plans to bring any of the constant reduction back for functions like `range`? If so can you link me to the relevant issues? Thanks!

---

edit: I band-aid fixed this on my end by changing the `REGISTER_FUNC()` for range to be `const`. This seems to restore the behaviours that this PR deleted.
